src/backend/dx11/src/dxgi.rs

```diff
@@ -140,30 +140,79 @@ fn enum_adapters1(
     }
 }
 
-fn get_adapter_desc(adapter: *mut dxgi::IDXGIAdapter, version: DxgiVersion) -> AdapterInfo {
+fn build_adapter_info(
+    description: &[u16],
+    vendor_id: u32,
+    device_id: u32,
+    flags: u32,
+    video_mem: usize,
+    device: &ComPtr<d3d11::ID3D11Device>,
+) -> AdapterInfo {
+    let mut features_options2: d3d11::D3D11_FEATURE_DATA_D3D11_OPTIONS2 = unsafe { mem::zeroed() };
+
+    let device_name = {
+        let len = description.iter().take_while(|&&c| c != 0).count();
+        let name = <OsString as OsStringExt>::from_wide(&description[..len]);
+        name.to_string_lossy().into_owned()
+    };
+
+    AdapterInfo {
+        name: device_name,
+        vendor: vendor_id as usize,
+        device: device_id as usize,
+        device_type: if (flags & dxgi::DXGI_ADAPTER_FLAG_SOFTWARE) != 0 {
+            DeviceType::VirtualGpu
+        } else if winerror::SUCCEEDED(unsafe {
+            device.CheckFeatureSupport(
+                d3d11::D3D11_FEATURE_D3D11_OPTIONS2,
+                &mut features_options2 as *mut _ as *mut _,
+                mem::size_of::<d3d11::D3D11_FEATURE_DATA_D3D11_OPTIONS2>() as _,
+            )
+        }) {
+            if features_options2.UnifiedMemoryArchitecture == 1 {
+                DeviceType::IntegratedGpu
+            } else {
+                DeviceType::DiscreteGpu
+            }
+        } else if video_mem > 512_000_000 {
```

Comment:
This check should be inverted. If we don't know if it's UMA or not, we should assume <= 512 is integrated.

In point of fact, I don't think we need the vendor based stuff at all. If none of these other checks succeed, we're Discrete.